locator/topology.cc

```diff
@@ -25,81 +25,264 @@ thread_local const endpoint_dc_rack endpoint_dc_rack::default_location = {
+        add_node(make_lw_shared<node>(cfg.local_host_id, cfg.local_endpoint, cfg.local_dc_rack, node::local::yes));
+    }
+}
+
+topology::topology(topology&& o) noexcept
+    : _shard(o._shard)
+    , _all_nodes(std::move(o._all_nodes))
+    , _local_node(std::move(o._local_node))
+    , _nodes_by_host_id(std::move(o._nodes_by_host_id))
+    , _nodes_by_endpoint(std::move(o._nodes_by_endpoint))
+    , _dc_nodes(std::move(o._dc_nodes))
+    , _dc_rack_nodes(std::move(o._dc_rack_nodes))
+    , _dc_endpoints(std::move(o._dc_endpoints))
+    , _dc_racks(std::move(o._dc_racks))
+    , _sort_by_proximity(o._sort_by_proximity)
+    , _datacenters(std::move(o._datacenters))
 {
-    update_endpoint(utils::fb_utilities::get_broadcast_address(), cfg.local_dc_rack);
+    assert(_shard == this_shard_id());
+    tlogger.trace("topology[{}]: move from [{}]", fmt::ptr(this), fmt::ptr(&o));
 }
 
 future<topology> topology::clone_gently() const {
     topology ret;
-    ret._dc_endpoints.reserve(_dc_endpoints.size());
-    for (const auto& p : _dc_endpoints) {
-        ret._dc_endpoints.emplace(p);
-    }
-    co_await coroutine::maybe_yield();
-    ret._dc_racks.reserve(_dc_racks.size());
-    for (const auto& [dc, rack_endpoints] : _dc_racks) {
-        ret._dc_racks[dc].reserve(rack_endpoints.size());
-        for (const auto& p : rack_endpoints) {
-            ret._dc_racks[dc].emplace(p);
+    if (this_shard_id() == _shard) {
+        tlogger.debug("topology[{}]: clone_gently to {} on same shard", fmt::ptr(this), fmt::ptr(&ret));
+        ret._all_nodes = _all_nodes;
+        co_await coroutine::maybe_yield();
+        ret._local_node = _local_node;
+        ret._nodes_by_host_id = _nodes_by_host_id;
+        ret._nodes_by_endpoint = _nodes_by_endpoint;
+        ret._dc_nodes = _dc_nodes;
+        ret._dc_rack_nodes = _dc_rack_nodes;
+        ret._dc_endpoints = _dc_endpoints;
+        ret._dc_racks = _dc_racks;
+        ret._datacenters = _datacenters;
+    } else {
+        tlogger.debug("topology[{}]: clone_gently to {} from shard {}", fmt::ptr(this), fmt::ptr(&ret), _shard);
+        for (const auto& n : _all_nodes) {
+            ret.add_node(make_lw_shared<node>(*n));
+            co_await coroutine::maybe_yield();
+        }
+        // local node may be detached for _all_nodes
+        // if it was decommissioned.
+        if (!ret._local_node && _local_node) {
+            ret.add_node(make_lw_shared<node>(*_local_node));
         }
     }
     co_await coroutine::maybe_yield();
-    ret._current_locations.reserve(_current_locations.size());
-    for (const auto& p : _current_locations) {
-        ret._current_locations.emplace(p);
-    }
-    co_await coroutine::maybe_yield();
-    ret._datacenters = _datacenters;
     ret._sort_by_proximity = _sort_by_proximity;
     co_return ret;
 }
 
-void topology::update_endpoint(const inet_address& ep, endpoint_dc_rack dr)
-{
-    auto current = _current_locations.find(ep);
```

Comment:
nit: adding the new code first, then removing the old one in a follow-up commit (where possible) would have made for a better review experience.

Reply:
I have the opposite experience, when a function replaces another, I like to compare the two versions.